### PR TITLE
8345052: Harden StampedLock

### PR DESCRIPTION
**Pull request by DougLea**
This applies the same kinds of recovery on OOME (by spinning) and JVM Exceptions (by cancelling) seen in AQS.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345052](https://bugs.openjdk.org/browse/JDK-8345052): Harden StampedLock (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Viktor Klang](https://openjdk.org/census#vklang) (@viktorklang-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22396/head:pull/22396` \
`$ git checkout pull/22396`

Update a local copy of the PR: \
`$ git checkout pull/22396` \
`$ git pull https://git.openjdk.org/jdk.git pull/22396/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22396`

View PR using the GUI difftool: \
`$ git pr show -t 22396`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22396.diff">https://git.openjdk.org/jdk/pull/22396.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22396#issuecomment-2501626697)
</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/concurrent/locks/StampedLock.java [hunk 1]
```diff
@@ -1244,21 +1256,25 @@ else if (!casTail(t, node))
...
+                        LockSupport.parkNanos(this, nanos);
+                } catch (Error | RuntimeException ex) {
+                    cancelAcquire(node);
+                    throw ex;
+                }
                 node.clearStatus();
                 if ((interrupted |= Thread.interrupted()) && interruptible)
                     break;
-            }
+                spins = postSpins = (byte)((postSpins << 1) | 1);
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/locks/StampedLock.java:1272]: @DougLea Any specific reason to move the updates to `spins` and `postSpins` to after the park?

Comment (DougLea) [reply src/java.base/share/classes/java/util/concurrent/locks/StampedLock.java:1272]: Just to improve clarity after the exceptional and break cases

Comment (viktorklang-ora) [reply src/java.base/share/classes/java/util/concurrent/locks/StampedLock.java:1272]: Makes sense—thanks

**Comment by AlanBateman** [comment]
Viktor is repeat testing the test as OOME tests have historically been problematic to be stable in all configs.

The changes to StampedLock looks okay to me. A minor comment on naming is that it feels like spinLockOnOOME is really acquireOnOOME or spinAcquireOnOOME but it's all internal and doesn't matter.

**Review by AlanBateman — APPROVED** [review]

**Review by viktorklang-ora — APPROVED** [review]
I ran a 10k repetition run of the test in this PR and it's all green.

I'm running some final tests just to be on the safe side, and I'll let you know if it's safe to integrate.

**Comment by viktorklang-ora** [comment]
@DougLea Looks good, please go ahead and integrate. 👍

**Pull request closed without merging**